### DISCUSSION TITLE
Different behaviour of nested models between amd64 and arm64

### DISCUSSION THREAD

**Discussion by dfioravanti** [pydantic/pydantic#4779] [Question]
Hey everyone, this one is quite a weird bug. I have a nested model which is more or less 
```
class A(BaseModel):
   foo: bar

IdsToA: Dict[str, A]

class B(BaseModel):
   ids_to_a : IdsToA
```
as I am converting an existing codebase to pydantic I need to export this to a dict which I do with 
```
b = b(<whatever>)
_dict = b.dict()
```
then I have a bunch of tests that I run with pytest. All of this works like a charm when I test is locally or when I build a docker image using `python:3.10-slim` with `linux/arm64` as platform. But it fails when I run it on gitlab and/or locally `linux/amd64` as platform. Then from the errors it looks like the elements of `ids_to_a` are not exported to dictionary but instead are still of class `A`. Any clue? I tried every workaround I can think off (exporting to json and reload it, etc) but nothing works

**Comment by samuelcolvin** [comment]
What do you get from `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`?

**Reply by dfioravanti** [reply]
On my machine 

```
             pydantic version: 1.10.2
            pydantic compiled: True
                 install path: <removed>
               python version: 3.10.8 (main, Oct 13 2022, 09:48:40) [Clang 14.0.0 (clang-1400.0.29.102)]
                     platform: macOS-12.6.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

on docker 

```
             pydantic version: 1.10.2
            pydantic compiled: True
                 install path: /usr/local/lib/python3.10/site-packages/pydantic
               python version: 3.10.8 (main, Nov 15 2022, 21:06:05) [GCC 10.2.1 20210110]
                     platform: Linux-5.15.49-linuxkit-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
  ```

**Reply by samuelcolvin** [reply]
Humm, I don't really know what to suggest. Without a reproducible example it's hard to help further.

I thought it might be that pydantic was not compiled in one of your environments, or you were using different versions, but that's not the case.

**Reply by dfioravanti** [reply]
Yea I understand, I will try to come up with a reproducible example that I can share. Thanks for the help

**Comment by dfioravanti** [comment]
In case some other poor soul bumps into this problem while migrating a codebase, I managed to find a workaround. Just create a custom encoder like
```py
class JSONResultEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, BaseModel):
            return o.dict(by_alias=True, exclude_none=True)

        return json.JSONEncoder.default(self, o)
```
then define 
```py
def custom_dumps(values, *, default):
    return JSONResultEncoder().encode(values)
```
and set 
```py
    class Config:
        json_dumps = custom_dumps
```
and then just call
```py
 json = self.json(by_alias=True, exclude_none=True)
 _dict = orjson.loads(json)
```
yep, it is incredibly silly but it works.

(updated to highlight python)

**Reply by samuelcolvin** [reply]
You shouldn't need a custom encoder. `model.json()` should already convert models to dicts.

If it's not and you're getting an error, please can you post it here.

Also, we have [`json_encoder`](https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders), if you really need to customise that.

**Reply by dfioravanti** [reply]
Yes, I know both exist but they do not work. I will try to provide a minimum working example where it break down. But thanks for the suggestion and for this amazing library =)
